### PR TITLE
#162170711 Enable Social Login via Google, Facebook and Twitter

### PR DESCRIPTION
**Pull request by akhilome**
#### What does this PR do?

Enable social authentication

#### Description of Task to be completed?

- install passport, other necessary plugins and setup appropriately
- implement social authentication using facebook oauth
- implement social authentication using google oauth
- implement social authentication using twitter oauth
- write tests for implemented endpoints
- write documentation for changes

#### How should this be manually tested?

- Clone the repo
- `cd` into `balder-ah-backend`
- rename `.env.sample` and fill out necessary details
- install packages by running `npm i`
- start the local server by running `npm run dev:start`
- hit any of the following endpoints
  * `localhost:3000/api/auth/google/`
  * `localhost:3000/api/auth/twitter/`
  * `localhost:3000/api/auth/facebook/`

#### Any background context you want to provide?

N/A

#### What are the relevant pivotal tracker stories?

[#162170711](https://www.pivotaltracker.com/story/show/162170711)

#### Screenshots (if appropriate)

image

image

#### Questions:

N/A

### CODE REVIEW

File: src/controllers/socialAuth.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+   * @static
+   *
+   * @param {object} request - HTTP request object
+   * @param {object} response - HTTP response object
+   *
+   * @returns {object} - object containing a generated token
+   * @memberof SocialAuthController
+   */
+  static getToken(request, response) {
+    const { id, username, newUser } = request.user;
```
Comment (frostyblok) [inline src/controllers/socialAuth.js]: Hi Kizito, nice job. Please I would like to know how you came about newUser?

Comment (akhilome) [reply src/controllers/socialAuth.js]: @frostyblok, admittedly, that variable name is not descriptive enough. `newUser` above signifies whether or not the user was created.

I'll make the variable name more descriptive. 

Thanks for the feedback. 🙏🏾

Comment (frostyblok) [reply src/controllers/socialAuth.js]: It is clear now. Good job, Kizito.

**Review by frostyblok — APPROVED** [review]

**Comment by omokaro-faith** [comment]
@akhilome Conflicts 😅😅😅😅😅😅😅😅

**Comment by akhilome** [comment]
> @akhilome Conflicts 😅😅😅😅😅😅😅😅

@omokaro-faith Fixed. 😊

**Review by marcdomain — APPROVED** [review]
LGTM!

File: src/server.js [hunk 2]
```diff
@@ -1,18 +1,27 @@
...
+} from './routes';
 import registerMiddlewares from './middlewares';
+import './services/passport';
 
 const app = express();
 const PORT = process.env.PORT || 3000;
 
 registerMiddlewares(app);
 
+app.use(passport.initialize());
```
Comment (luqmanoop) [inline src/server.js:18]: You've done a great job implementing this feature. However, I feel there's no need to clutter the `server.js` file with different middleware registrations. We have a `middlwares/index.js` file which serves as a single source of truth for all middlewares the app uses. Here on out, please, endeavor to register new middlewares in the `middlewares/index.js` file.  Cheers 👍

**Review by luqmanoop — APPROVED** [review]
LGTM! 👍

File: src/server.js (hunk 2 above)
Comment (akhilome) [reply src/server.js:18]: Thanks for the feedback, Luqman. 👍

File: src/controllers/socialAuthController.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+import { generateToken } from '../middlewares/authentication';
+import models from '../db/models';
+
+const { User } = models;
+
+/**
+ * @description Social authentication controller
+ * @class SocialAuthController
+ */
+class SocialAuthController {
```
Comment (okoroemeka) [inline src/controllers/socialAuthController.js:11]: Nice work Kizito, but however I think we could change the class name to SocilaAuthentication without the adding the controller part.

File: src/controllers/socialAuthController.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+      where: {
+        $or: [{ email: emails[0].value }, { googleid: id }]
+      },
+      defaults: {
+        username: `${name.givenName}${id.slice(0, 6)}`.toLowerCase(),
+        email: emails[0].value,
+        googleid: id,
+        password: crypto.randomBytes(20).toString('hex')
+      }
+    }).spread((user, created) => {
```
Comment (okoroemeka) [inline src/controllers/socialAuthController.js:43]: What does `.spread` do?

**Pull request merged**